services/src/main/scala/cromwell/services/instrumentation/impl/stackdriver/StackdriverInstrumentationServiceActor.scala

```diff
@@ -0,0 +1,183 @@
+    val metricObj = StackdriverMetric(bucket.toStackdriverString, metricKind)
+
+    if (metricsMap.contains(metricObj)) {
+      val valueList = metricsMap(metricObj)
+      metricsMap += metricObj ->  valueList.::(metricValue)
+    }
+    else metricsMap += metricObj -> List(metricValue)
+  }
+
+
+  private def sendMetricData(): Unit = {
+    metricsMap.foreach { case (key, value: List[Double]) =>
+      val dataPointListSum = value.sum
+      val dataPoint = key.kind match {
+        case StackdriverGauge => dataPointListSum / value.length
+        case StackdriverCumulative => dataPointListSum
+      }
+      writeMetrics(key, dataPoint)
+    }
+
+    metricsMap = Map.empty[StackdriverMetric, List[Double]]
+  }
+
+
+  private def timeInterval(metricKind: StackdriverMetricKind): TimeInterval = {
+     metricKind match {
+       case StackdriverGauge => TimeInterval.newBuilder.setEndTime(Timestamps.fromMillis(System.currentTimeMillis)).build
+       case StackdriverCumulative => TimeInterval.newBuilder.setStartTime(ActorCreationTime).setEndTime(Timestamps.fromMillis(System.currentTimeMillis)).build
+     }
+  }
+
+
+  private def createTimeSeries(metricKind: StackdriverMetricKind, metric: Metric, resource: MonitoredResource, dataPointList: util.List[Point]): TimeSeries = {
+    metricKind match {
+      case StackdriverGauge => TimeSeries.newBuilder.setMetric(metric).setResource(resource).addAllPoints(dataPointList).build
+      case StackdriverCumulative => TimeSeries.newBuilder.setMetric(metric).setResource(resource).setMetricKind(MetricDescriptor.MetricKind.CUMULATIVE).addAllPoints(dataPointList).build
+    }
+  }
+
+
+  private def createMetricBuilder(metricName: String): Metric = {
+    metricLabelsMap match {
+      case map: Map[String, String] if map.isEmpty => Metric.newBuilder.setType(s"$CustomMetricDomain/$metricName").build
+      case map: Map[String, String] if map.nonEmpty => Metric.newBuilder.setType(s"$CustomMetricDomain/$metricName").putAllLabels(map.asJava).build
+    }
+  }
+
+
+  private def writeMetrics(metricObj: StackdriverMetric, value: Double): Unit = {
+    // Prepares an individual data point
+    val interval = timeInterval(metricObj.kind)
+    val pointValue = TypedValue.newBuilder().setDoubleValue(value).build()
+    val dataPoint: Point = Point.newBuilder.setInterval(interval).setValue(pointValue).build
+    val dataPointList: List[Point] = List[Point](dataPoint)
+
+    // Prepares the metric descriptor
+    val metric: Metric = createMetricBuilder(metricObj.name)
+
+    // Prepares the time series request
+    val timeSeries = createTimeSeries(metricObj.kind, metric, monitoredResource, dataPointList.asJava)
+    val timeSeriesList = List[TimeSeries](timeSeries)
+
+    val timeSeriesRequest = CreateTimeSeriesRequest.newBuilder.setName(projectName.toString).addAllTimeSeries(timeSeriesList.asJava).build
+
+    // Writes time series data
+    metricServiceClient.createTimeSeries(timeSeriesRequest)
```

Comment:
TOL:

The javadoc for this method says:
```
   * @throws com.google.api.gax.rpc.ApiException if the remote call fails
```

If that happens then the actor will crash and - since this is probably using the default strategy - be restarted with its state reset to nothing (ie it would have no data in the metricMap).

That might be what you want (since it's the simplest answer and it's not the end of the world if it does happen) - but you might want to handle these exceptions more gracefully?

Reply:
That's a good point. I will handle the exception.